File: backend/app/schemas/ontology.py
from __future__ import annotations
from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel, Field


# --- ObjectType schemas ---


class ObjectTypeCreate(BaseModel):
    name: str = Field(..., max_length=100)
    icon: Optional[str] = None
    color: Optional[str] = None
    description: Optional[str] = None
    properties_schema: Optional[dict] = None


class ObjectTypeUpdate(BaseModel):
    name: Optional[str] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    description: Optional[str] = None
    properties_schema: Optional[dict] = None


class ObjectTypeResponse(BaseModel):
    id: str
    name: str
    icon: Optional[str]
    color: Optional[str]
    description: Optional[str]
    properties_schema: Optional[dict]
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Object (ObjectInstance) schemas ---


class ObjectCreate(BaseModel):
    type_id: str
    title: str = Field(..., max_length=500)
    properties: Optional[dict] = None
    status: Optional[str] = None
    severity: Optional[str] = None


class ObjectUpdate(BaseModel):
    title: Optional[str] = None
    properties: Optional[dict] = None
    status: Optional[str] = None
    severity: Optional[str] = None


class ObjectResponse(BaseModel):
    id: str
    type_id: str
    type_name: Optional[str] = None
    title: str
    properties: Optional[dict]
    status: Optional[str]
    severity: Optional[str]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ObjectListResponse(BaseModel):
    items: List[ObjectResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


# --- Link schemas ---


class LinkCreate(BaseModel):
    source_id: str
    target_id: str
    link_type: str = Field(..., max_length=100)
    properties: Optional[dict] = None


class LinkResponse(BaseModel):
    id: str
    source_id: str
    target_id: str
    link_type: str
    properties: Optional[dict]
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Action schemas ---


class ActionCreate(BaseModel):
    name: str = Field(..., max_length=100)
    object_type_id: str
    action_type: Optional[str] = None
    config: Optional[dict] = None


class ActionResponse(BaseModel):
    id: str
    name: str
    object_type_id: str
    action_type: Optional[str]
    config: Optional[dict]
    created_at: datetime

    model_config = {"from_attributes": True}


class ActionExecuteRequest(BaseModel):
    object_id: str


# --- Graph schemas ---


class GraphNode(BaseModel):
    id: str
    type_id: str
    type_name: Optional[str] = None
    title: str
    status: Optional[str] = None
    severity: Optional[str] = None


class GraphEdge(BaseModel):
    id: str
    source_id: str
    target_id: str
    link_type: str


class GraphResponse(BaseModel):
    nodes: List[GraphNode]
    edges: List[GraphEdge]

File: frontend/src/types/ontology.ts
export type Severity = 'critical' | 'high' | 'medium' | 'low' | 'info';
export type ObjectStatus = 'active' | 'resolved' | 'mitigated' | 'false_positive' | 'risk_accepted';

export interface PropertySchema {
  key: string;
  label: string;
  type: 'string' | 'number' | 'boolean' | 'date' | 'enum' | 'json';
  required: boolean;
  enumValues?: string[];
}

export interface ObjectTypeDefinition {
  id: string;
  name: string;
  icon: string;
  color: string;
  description: string;
  propertiesSchema: PropertySchema[];
  createdAt: string;
}

export interface OntologyObject {
  id: string;
  typeId: string;
  typeName: string;
  title: string;
  properties: Record<string, unknown>;
  status: string;
  severity?: Severity;
  createdAt: string;
  updatedAt: string;
}

export interface ObjectLink {
  id: string;
  sourceId: string;
  targetId: string;
  linkType: string;
  properties?: Record<string, unknown>;
  createdAt: string;
}

export interface ObjectAction {
  id: string;
  name: string;
  objectTypeId: string;
  actionType: 'jira' | 'isolate' | 'remediate' | 'notify' | 'custom';
  config: Record<string, unknown>;
}

export interface GraphNode {
  id: string;
  label: string;
  type: string;
  color: string;
  properties?: Record<string, unknown>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  label: string;
}

export interface GraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}
